File: 0101-0200/0106.Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.js
//1ms

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
const build = (postorder, inorderMap, indexObj, start = 0, end = 0) => {
  if (start > end) return null;
  const rootVal = postorder[indexObj.value--];
  const root = new TreeNode(rootVal);
  const inorderI = inorderMap.get(rootVal);
  root.right = build(postorder, inorderMap, indexObj, inorderI + 1, end);
  root.left = build(postorder, inorderMap, indexObj, start, inorderI - 1);
  return root;
};

var buildTree = function (inorder, postorder) {
  const inorderMap = new Map();
  for (let i = 0; i < inorder.length; i++) inorderMap.set(inorder[i], i);
  const indexObj = { value: postorder.length - 1 };
  return build(postorder, inorderMap, indexObj, 0, inorder.length - 1);
};

File: 0101-0200/0106.Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.rs
//0ms

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;

impl Solution {
    pub fn build_tree(inorder: Vec<i32>, postorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        fn helper(
            postorder: &[i32],
            inorder_map: &HashMap<i32, usize>,
            q: usize
        ) -> Option<Rc<RefCell<TreeNode>>> {
            let [rest @ .., val] = postorder else { return None; };
            let inorder_idx = inorder_map[val];
            let i = inorder_idx - q;
            Some(Rc::new(RefCell::new(TreeNode {
                val: *val,
                left: helper(&rest[..i], inorder_map, q),
                right: helper(&rest[i..], inorder_map, inorder_idx + 1),
            })))
        }

        helper(&postorder, &inorder.into_iter().enumerate().map(|(i, v)| (v, i)).collect(), 0)
    }
}
